### ISSUE TITLE
Change Theme at runtime

### ISSUE THREAD

**Issue by Springrbua** [vaadin/flow#5163]
**Problem:**
The theme is currently set using the `@Theme`-annotation and it does not seem to be possible to change it at runtime.

**Example use case:**
If you create a webshop, which could be used by many different clients, you probably want to match their Corporate Identity / Corporate Design.
One company might prefer the Lumo-Theme while others might prefer the Material-Theme.
One might even have custom `Theme`-Classes.
At the moment, you have to build one version (*.jar /*.war) for every theme.
In my opinion, it would be better to have a way to configure this behavior.

**Comment by appreciated** [comment]
> **Problem:**
The theme is currently set using the `@Theme`-annotation and it does not seem to be possible to change it at runtime.

Actually no a big deal:
```
@Override
protected void onAttach(AttachEvent attachEvent) {
    super.onAttach(attachEvent);
    getUI().get().getPage().executeJavaScript("document.documentElement.setAttribute(\"theme\",\"dark\")");
}
```

**Comment by Springrbua** [comment]
@appreciated this way you can only set the theme variant. But you can't switch from Lumo to Material I guess.
If you change the theme using `@Theme(Material.class)`, the imports (`@HtmlImport`) from `src` are actaully loaded form `theme/material`.
This logic probably won't be triggered by setting the `theme`-Attribute.

**Comment by web-padawan** [comment]
It is not possible to switch from Lumo to Material for the web components either, not only because of how it is implemented in Flow. Importing the theme adds custom CSS properties and these are processed for polyfilled browsers like IE11, so there is no API in the ShadyCSS to "opt-out" them.

**Comment by Springrbua** [comment]
@web-padawan I understand, that it is not possible to switch the theme on-the-fly, but I guess it should be possible to activate it after a page refresh?

**Comment by pleku** [comment]
Yes @Springrbua you'd need to do a page refresh to get the theme changes applied.

To allow dynamically changing the theme, you should use the `@NoTheme` annotation and then determine the used theme using something else (request parameters ?) and decide the included themed files during runtime by using a [dependency filter]https://vaadin.com/docs/v12/flow/advanced/tutorial-dependency-filter.html.

While I would want us to do some changes for how the themes are handled and selected in Flow (using Java), we will not probably have time for this before summer, which means that it would be earliest available in Vaadin 15 (if we decide to do this).

**Comment by Springrbua** [comment]
@pleku thank you for your informations.
Fortunately I don't need this feature yet, because we are still early in development but I am glad to hear that there is a solution for it.

**Comment by Legioth** [comment]
As long as the differences between the different themes are small enough, you could potentially also do the changes only based setting different CSS variables for Lumo. This can be done on-the-fly using https://github.com/OlliTietavainenVaadin/cssvariablesetter.

**Comment by Springrbua** [comment]
@Legioth unfortunately this is not enough. The `vaadin-button` for example uses completely different styles for material and lumo theme. Those styles are based on two different html-files and they also use different CSS-Variables + different Theme-Variants.
Changing the color-palette etc. at runtime should not be a big problem though.

**Comment by Springrbua** [comment]
@pleku I found a way to change the Theme at runtime by creating a custom `UI`-class and overriding the `getThemeFor`-method:
```
public class CustomUI extends UI {
    @Override
    public Optional<ThemeDefinition> getThemeFor(Class<?> navigationTarget, String path) {
        User user = getSession().getAttribute(User.class);
        return Optional.of(new ThemeDefinition(user.getThemeClass(), user.getThemeVariant()));
    }
}
```

This solution does not seem to be too hacky to me. What do you say?

**Comment by Legioth** [comment]
It does indeed seem like  `UI.getThemeFor` should work for selecting which theme to use when a page is loaded. You'd still have to manually ensure that all relevant theme files are present in the production bundle.

**Comment by Springrbua** [comment]
Of course, all needed files need to be in the production bundle and a page-refresh needs to be triggered.
But other then that, it seems straightforward to me and might even become the official solution for this problem.
If you agree with me, this issue can be closed but a documentation should be added to the [Theming-Section](https://vaadin.com/docs/v13/flow/theme/theming-overview.html).

**Comment by gokhanturkeri** [comment]
You can add combobox to your menu bar or where ever you want.

```
ComboBox<Byte> comboBoxTheme = new ComboBox<>();
comboBoxTheme.addValueChangeListener(listener -> {
            if (listener.getValue() != null) {
                if (listener.getValue() == (byte) 0) {
                    UI.getCurrent().getElement().setAttribute("theme", "dark");
                } else
                    UI.getCurrent().getElement().setAttribute("theme", "light");
            }
        });
```

**Comment by Springrbua** [comment]
@gokhanturkeri with your solution you can only change theme-variants, but not the whole theme. I need to be able to switch from `Lumo` to `Material` dynamically.
This is only possible with the [solution posted above](https://github.com/vaadin/flow/issues/5163#issuecomment-483237172).

**Comment by Springrbua** [comment]
I just wanted to let you know, that my solution/workaround doesn't work since Vaadin 14.  
This seems to make sense, since the frontend files are now bundled and the bundle will probably only contain the code for one theme.

**Comment by oliveryasuna** [comment]
This would be a great feature. Even if a refresh is necessary, you should be able to change the theme at runtime.

**Comment by mrts** [comment]
@probert94 
> Fortunately I don't need this feature yet, because we are still early in development but I am glad to hear that there is a solution for it.

How did you solve this in your product?

**Comment by probert94** [comment]
I decided to support Lumo theme only.
The reason for this is that the two themes differ quite a lot so it is hard to create a layout that works for both, without unneccessary scrollbars etc.

**Comment by oliveryasuna** [comment]
@Legioth Does Vaadin include all themes in the `themes` directories at production build time? If not, @gokhanturkeri's workaround will not work.

**Comment by oliveryasuna** [comment]
This just became relevant for my company. We are developing a whitelabel app and our client wants a single production instance (non-negotiable). We could use an attribute on the `body` tag, but that's messy and it exposes other variants to wrong customers. Sure, we could separate our CSS files and use some security mechanism to prevent loading particular ones based on the customer type, but that's also messy.

Since I want to do this in a maintainable and extendable manner, I am basically rewriting application-theme-plugin and overhauling the server-side theme system.

**Comment by rolfsmeds** [comment]
In the vast majority of cases, you probably don't want to switch _all_ the CSS in your UI, but rather have a shared foundation on top of which you then load a dynamically selected stylesheet (let's call that a "subtheme") that sets colors and other styles.

If it's a reasonably small set of styles, with a small set of hardcoded alternatives, you could probably just have them all baked into your theme, and switch between them using classnames or something, as described here: https://vaadin.com/docs/latest/styling/lumo/lumo-style-properties

If each "subtheme" is rather large, and you don't want to load all of them into the browser for everybody, or if you need to be able to add new subthemes on the fly without building a new version of your app, you could load those additional stylesheets dynamically, as described here: https://vaadin.com/docs/latest/styling/lumo/lumo-style-properties

Neither case supports switching between Lumo and Material (or, as will be the options in V25, Lumo and the new Aura theme), but as @probert94 already [mentioned](https://github.com/vaadin/flow/issues/5163#issuecomment-1422044326) above, it's usually not so easy to make your UI work well with two completely different themes, so maybe that's not what you really want to do anyway.

> the two themes differ quite a lot so it is hard to create a layout that works for both, without unneccessary scrollbars etc.

**Comment by Artur-** [comment]
With Vaadin 25 the whole idea is not to have a `@Theme` concept anymore but instead load standard CSS files. Also with the removal of the need for shadow root style injection and the removal of the Material theme, a lot of things are simplified.

You can today already load CSS files at runtime using `getUI().getPage().addStyleSheet("my.css")` so if you want to implement the "customize per user or company" case you would have
1. A standard CSS file containing the "base" for your application - always loaded using e.g. `@CssImport`
2. A customer or user specific style sheet in `src/main/resources/META-INF/resources/company1.css` that you would load as `getUI().getPage().addStyleSheet("company1.css")` when that user logs in, or when the hostname matches or however

If you would want to decide the full theme based on some parameter, you would change the `@CssImport` to use `addStyleSheet()` also.

If/once #22305 is merged, you can also remove CSS files during runtime using the return value from `addStyleSheet`, which means that you are able to unload `company1.css` and load `company2.css` instead, if you need to. Or unload `my-full-theme1.css` and load `my-full-theme2.css`.

Is there anything else to do in this issue or should it be closed after #22305?

**Comment by probert94** [comment]
In Vaadin 24.X a theme can not only change the style but also the template (actually the whole js logic) of a component.
If thats not the case anymore with Vaadin 25, then the issue can be closed in my opion as dynamic CSS themes are already possible.

**Comment by Legioth** [comment]
> In Vaadin 24.X a theme can not only change the style but also the template (actually the whole js logic) of a component.

I'm not sure what this refers to. Can you please elaborate?

**Comment by Artur-** [comment]
Different JS imports for material and lumo, most likely

**Comment by Legioth** [comment]
I guess that mechanism can indeed be used to make functional changes as well though my understanding is that it's only used to load different styles the built-in Vaadin components.

**Comment by probert94** [comment]
@Legioth @Artur- yes thats exactly what I meant, until now the js-files under theme/material and theme/lumo were used and could have differen styles, templates and even logic. While the logic should always be the same for different themes, changing the template could make sense.

**Comment by web-padawan** [comment]
> the js-files under theme/material and theme/lumo were used and could have differen styles, templates and even logic.

Just to clarify, templates and logic didn't change across theme/lumo and theme/material. Only styles were different.

**Comment by mrts** [comment]
Just a reminder that Vaadin 8 had a wonderfully [simple API](https://vaadin.com/docs/v8/framework/articles/ChangingThemeOnTheFly) for changing the theme at runtime:

```java
UI.setTheme("theme name");
```

That's the functionality we are looking for.

**Comment by Artur-** [comment]
With the linked PR, that would be something like
```
private Registration activeThemeRemover;
...
Page page = getUI().get().getPage();
activeThemeRemover = page.addStyleSheet("theme1.css");
...
activeThemeRemover.remove();
activeThemeRemover = page.addStyleSheet("theme2.css");
```

**Comment by Artur-** [comment]
Closing this as #22305 is included in 25.0

**Comment by mshabarov** [comment]
Docs update needed https://github.com/vaadin/docs/issues/5009.
